php/php-rr-sdk.md

```diff
@@ -0,0 +1,680 @@
+    });       
+
+    yield new WaitSignal('name');
+}
+```
+
+### Logging
+The workflow logging does not differ from any other parts of the application, the only exception is the need to exclude
+logs triggered while workflow is replaying. We propose to use mocked PSR-3 logger for this purposes:
+
+```php
+$logger = new NonReplayingLogger(Workflow::ctx(), $parentLogger); // it is possible to omit ::ctx()
+```
+
+### Sessions
+Sessions can be created as sub-activity group, using similar approach as `withTimeout` and other methods. However,
+the explicit yield required. 
+
+```php
+public function run(string $input): string
+{
+    $session = new Session([
+        Session::START_TO_FINISH_TIMEOUT => 300,
+        // ... 
+    ]);
+    
+    $a1 = $session->executeActivity(new ExecuteActivity('activity1'));
+    $a2 = $session->executeActivity(new ExecuteActivity('activity2'));
+    
+    $result1 = yield $a1;
+    $result2 = yield $a2; 
+
+    // required    
+    yield $session->close();
+    
+    return $result;
+}
+```
+
+### Deterministic Time
+Workflows must avoid calling SPL functions `time()` and `date()`. Context method must be used instead:
+
+```php
+Workflow::now(); //DateTimeImmutable object.
+```
+
+### Side Effects
+Similar to Golang SDK the side effects can be registered using yield call:
+
+```php
+$result = Workflow::sideEffect(function(){
+    return mt_rand(0, 1000);
+});
+```
+
+> Similar to other SDKs, PHP version **does not** guarantee that sideEffect won't be called again during the replay. 
+
+### Error Handling
+Like in Java version of Temporal SDK the workflow errors can be delivered in form of exceptions triggered on `yield`:
+
+```php
+public function run(string $input): string
+{
+    try {
+        return yield $this->activities->doSomething($input);
+    } catch (ActivityException $e) {
+        // get related activity ID
+        $this->logger->warning(
+            "Activity %s failed with %s",
+            $e->getActivityID(),
+            $e->getMessage()
+        );       
+    }
+}
+``` 
+
+> The exception tree must follow similar design as Go and Java SDK.
+
+### Termination and Cancel Requests
+The cancel requests will be supplied to the workflow in a form of exceptions on `yield` call.
+Similar to activity errors workflow can handle these exceptions and perform a proper rollback.
+
+```php
+public function run(string $input): string
+{
+    $step = 0;
+    try {
+        $resultA = yield $this->activities->doA($input);
+        $step++;
+        
+        $resultB = yield $this->activities->doB($resultA);
+        $step++;
+
+        return yield $this->activities->doC($resultB);
+    } catch (CancellationException $e) {
+        switch ($step) {
+            case 2:
+                yield $this->activities->rollbackB($resultB);
+                // no break
+            case 1:
+                yield $this->activities->rollbackA($resultA);
+        }
+    
+        throw $e;   
+    }
+}
+```
```

Comment:
There is a need to give workflow authors an ability to cancel parts of a workflow. This would require introduction of some form of CancellationScope. See the [Java SDK version of it](https://github.com/temporalio/java-sdk/blob/112dd5d31cd7e184730e471f3908faa71644da3c/src/main/java/io/temporal/workflow/Workflow.java#L580).

Reply:
Hmmm, this one I need more clarification on.